target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c

```diff
@@ -2396,23 +2397,34 @@ static void rtl930x_led_init(struct rtl838x_switch_priv *priv)
 		return;
 	}
 
+	for (int i = 0; i < 4; i++) {
+		const __be32 *led_set;
+		char set_name[9];
+		u32 setlen;
+		u32 v;
+
+		sprintf(set_name, "led_set%d", i);
+		led_set = of_get_property(node, set_name, &setlen);
+		if (!led_set || setlen != 16)
+			break;
+		led_sets = i;
+		v = be32_to_cpup(led_set) << 16 | be32_to_cpup(led_set + 1);
+		sw_w32(v, RTL930X_LED_SET0_0_CTRL - 4 - i * 8);
+		v = be32_to_cpup(led_set + 2) << 16 | be32_to_cpup(led_set + 3);
+		sw_w32(v, RTL930X_LED_SET0_0_CTRL - i * 8);
```

Comment:
Some mental notes:
Index nr 3 is written to the lowest bits of `SETi_0_CTRL`, index nr 2 to the highest bits, index nr 1 to the lowest of `SETi_1_CTRL`, index nr 0 the hightest bits. At first glance, this means that structure is `led_setX = <LED3 LED2 LED1 LED0>` (which is a bit backwards, but OK).

You might be able use the `LEDx` values to determine how many LEDs a port is configured to use, but I see that for the XGS1250-10 the `LED2` value is set to 0xffff, i.e. matching all speeds and other conditions, which is also strange.

Reply:
LED setup reworked, check latest commits